tools/clang/unittests/SPIRV/LibTestUtils.cpp

```diff
@@ -0,0 +1,228 @@
+  if (tokens.size() < 3 || tokens[1].find("RUN") == std::string::npos ||
+      tokens[2].find("%dxc") == std::string::npos) {
+    fprintf(stderr, "The only supported format is: \"// RUN: %%dxc -T "
+                    "<profile> -E <entry>\"\n");
+    return false;
+  }
+
+  std::ostringstream rest;
+  for (uint32_t i = 3; i < tokens.size(); ++i) {
+    if (tokens[i] == "-T" && (++i) < tokens.size()) {
+      *targetProfile = tokens[i];
+    } else if (tokens[i] == "-E" && (++i) < tokens.size()) {
+      *entryPoint = tokens[i];
+    } else if (tokens[i].substr(0, 17) == "-fspv-target-env=") {
+      std::string targetEnvStr = tokens[i].substr(17);
+      if (targetEnvStr == "vulkan1.0")
+        *targetEnv = SPV_ENV_VULKAN_1_0;
+      else if (targetEnvStr == "vulkan1.1")
+        *targetEnv = SPV_ENV_VULKAN_1_1;
+      else if (targetEnvStr == "vulkan1.1spirv1.4")
+        *targetEnv = SPV_ENV_VULKAN_1_1_SPIRV_1_4;
+      else if (targetEnvStr == "vulkan1.2")
+        *targetEnv = SPV_ENV_VULKAN_1_2;
+      else if (targetEnvStr == "vulkan1.3")
+        *targetEnv = SPV_ENV_UNIVERSAL_1_6;
+      else if (targetEnvStr == "universal1.5")
+        *targetEnv = SPV_ENV_UNIVERSAL_1_5;
+      else {
+        fprintf(stderr, "Error: Found unknown target environment.\n");
+        return false;
+      }
+      // Also push target environment to restArgs so that it gets passed to the
+      // compile command.
+      restArgs->push_back(tokens[i]);
+    } else {
+      restArgs->push_back(tokens[i]);
+    }
+  }
+
+  if (targetProfile->empty()) {
+    fprintf(stderr, "Error: Missing target profile argument (-T).\n");
+    return false;
+  }
+  // lib_6_* profile doesn't need an entryPoint
+  if (targetProfile->c_str()[0] != 'l' && entryPoint->empty()) {
+    fprintf(stderr, "Error: Missing entry point argument (-E).\n");
+    return false;
+  }
+  return true;
+}
+
+void convertIDxcBlobToUint32(const CComPtr<IDxcBlob> &blob,
+                             std::vector<uint32_t> *binaryWords) {
+  size_t num32BitWords = (blob->GetBufferSize() + 3) / 4;
+  std::string binaryStr((char *)blob->GetBufferPointer(),
+                        blob->GetBufferSize());
+  binaryStr.resize(num32BitWords * 4, 0);
+  binaryWords->resize(num32BitWords, 0);
+  memcpy(binaryWords->data(), binaryStr.data(), binaryStr.size());
+}
+
+bool compileCodeWithSpirvGeneration(const llvm::StringRef code,
+                                    const llvm::StringRef entryPoint,
+                                    const llvm::StringRef targetProfile,
+                                    const std::vector<std::string> &restArgs,
+                                    std::vector<uint32_t> *generatedBinary,
+                                    std::string *errorMessages) {
+  std::wstring srcFile(L"");
+  std::wstring entry(entryPoint.begin(), entryPoint.end());
+  std::wstring profile(targetProfile.begin(), targetProfile.end());
+
+  std::vector<std::wstring> rest;
```

Comment:
Isn't it possible to use:
```cpp
std::vector<std::wstring> rest(restArgs.begin(), restArgs.end())
```
? Or is the wstring to string conversion causing issues?

Reply:
That does not compile because there is not constructor for std::wstring that takes an std::string, which is why `arg` is not passed directly to `emplace_back` below, but we use `begin` and `end`.